File: problem_solving_algorithms/easy/time_conversion.js
// Time Conversion
// https://www.hackerrank.com/challenges/time-conversion/problem

function timeConversion(s) {
    // filter and list all characters, removing ":"
    s = s.split("");
    var time_array = s.filter(function (el) {
        return !(el == ":");
    }) 

    // Create a new list in pairs of two
    var conversion_pairing = [];
    for (var i = 0; i < time_array.length; i+=2) {
        conversion_pairing.push(time_array[i].concat(time_array[i + 1]));
    }

    // If the end is AM, leave as is unless first hour is 12, then subtract 12 from first hour and append a "0"
    if (conversion_pairing[conversion_pairing.length - 1] === "AM") {
        if (conversion_pairing[0] === "12") {
            var converted_hour = parseInt(conversion_pairing[0]) - 12;
            conversion_pairing[0] = converted_hour.toString().concat("0");
            return conversion_pairing.slice(0, 3).join(":");
        } else {
            return conversion_pairing.slice(0, 3).join(":")
        }
    } else {
        // if the end is PM, add 12 to the first hour unless it is 12, then leaves as is
        if (conversion_pairing[0] === "12") {
            return conversion_pairing.slice(0,3).join(":");
        } else {
            var converted_hour = parseInt(conversion_pairing[0]) + 12;
            conversion_pairing[0] = converted_hour.toString();
            return conversion_pairing.slice(0,3).join(":");
        }
    }
}

console.log(timeConversion("07:05:45AM"));
console.log(timeConversion("07:05:45PM"));
console.log(timeConversion("12:40:22AM"));
console.log(timeConversion("12:40:22PM"));

File: problem_solving_algorithms/easy/time_conversion.py
# Time Conversion
# https://www.hackerrank.com/challenges/time-conversion/problem

def timeConversion(s):
    # Remove the colons
    s = list(filter((":").__ne__, s))

    # Create a list of characters in pairs of two
    time_array = []
    for i in range(0, len(s), 2):
        time_array.append(s[i] + s[i+1])

    # if last item in time_array is AM or PM, convert
    if time_array[-1] == "AM":
        # if hour is 12, subtract 12
        if time_array[0] == "12":
            converted_hour = str(int(time_array[0]) - 12) + "0"
            time_array[0] = converted_hour
            return ":".join(time_array[0:3])
        # if hour is not 12, leave as is
        else:
            return ":".join(time_array[0:3])
    else:
        # convert the first hour by adding 12 if its not 12 (noon)
        if time_array[0] == "12":
            return ":".join(time_array[0:3])
        else:
            converted_hour = str(int(time_array[0]) + 12)

        # replace the previous hours with the converted hour
        time_array[0] = converted_hour
        return ":".join(time_array[0:3])
    
print(timeConversion("07:05:45AM"))
print(timeConversion("07:05:45PM"))
print(timeConversion("12:40:22AM"))
print(timeConversion("12:40:22PM"))